src/LeafMutator.ts

```diff
@@ -0,0 +1,47 @@
+import ActionCreator from './interfaces/ActionCreator';
+import ActionMessage from './interfaces/ActionMessage';
+import { getPrivateActionId } from './actionCreator';
+
+export type MutatorHandler<TState, TAction extends ActionMessage> = (
+    state: TState,
+    action: TAction
+) => TState | void;
+
+// Represents a mutator for a leaf node in the state tree
+export default class LeafMutator<TState> {
+    private handlers: { [actionId: string]: MutatorHandler<TState, ActionMessage> } = {};
+
+    constructor(private initialValue: TState) {}
+
+    getInitialValue() {
+        return this.initialValue;
+    }
+
+    handles<TAction extends ActionMessage>(
+        actionCreator: ActionCreator<TAction>,
+        handler: MutatorHandler<TState, TAction>
+    ) {
+        let actionId = getPrivateActionId(actionCreator);
+        if (this.handlers[actionId]) {
+            throw new Error('A mutator may not handle the same action twice.');
+        }
+
+        this.handlers[actionId] = handler;
+        return this;
+    }
+
+    handleAction(
+        currentState: TState,
+        actionMessage: ActionMessage,
+        replaceState: (newState: TState) => void
+    ) {
+        let actionId = getPrivateActionId(actionMessage);
+        let handler = this.handlers[actionId];
+        if (handler) {
+            let returnValue = handler(currentState, actionMessage);
+            if (returnValue !== undefined) {
```

Comment:
This is part of the reason I don't like the ability to return a replacement value, especially *in addition to* mutating the value.

1. It is totally valid to set a value to undefined. However, if you return `undefined`, then nothing happens (due to seeming black magic)
2. If you have a handler like so:
    ```typescript
    .handles(actionA, (state, action) =>
          state.someProperty = 'A';
      )
    ```
    This actually results in replacing the entire state with `'A'`, rather than merely mutating it. All because of the lack of curly braces.

Reply:
1. The point about `undefined` is inarguable and is, I think, a pitfall that is pretty rarely relevant but people will need to be aware of.  It's akin to the MobX gotcha around not being able to observe properties that don't exist.
2. Possible in pure JS, but TypeScript would flag this as an error because your return value would be of the wrong type.

My personal feeling is that these downsides aren't strong enough to clutter up the API with two different methods, but that would be the alternative.  I'd like to gather a few more opinions and then settle on which way to do it.

---
In reply to: [200790946](https://github.com/Microsoft/satcheljs/pull/100#discussion_r200790946) [](ancestors = 200790946)